apps/src/lib/ui/accounts/ManageLinkedAccounts.jsx

```diff
@@ -2,103 +2,129 @@ import React, {PropTypes} from 'react';
 import BootstrapButton from './BootstrapButton';
+import {connect} from 'react-redux';
+import {disconnect} from './manageLinkedAccountsRedux';
 
 const OAUTH_PROVIDERS = {
   GOOGLE: 'google_oauth2',
-  FACEBOOK: 'facebook',
-  CLEVER: 'clever',
   MICROSOFT: 'windowslive',
+  CLEVER: 'clever',
+  FACEBOOK: 'facebook',
 };
 export const ENCRYPTED = `*** ${i18n.encrypted()} ***`;
+const authOptionPropType = PropTypes.shape({
+  id: PropTypes.number.isRequired,
+  credentialType: PropTypes.string.isRequired,
+  email: PropTypes.string,
+  error: PropTypes.string,
+});
+const EMPTY_AUTH_OPTION = {
+  credentialType: '',
+  email: '',
+  error: '',
+};
 
-export default class ManageLinkedAccounts extends React.Component {
+class ManageLinkedAccounts extends React.Component {
   static propTypes = {
-    userType: PropTypes.string.isRequired,
-    authenticationOptions: PropTypes.arrayOf(PropTypes.shape({
-      id: PropTypes.number.isRequired,
-      credential_type: PropTypes.string.isRequired,
-      email: PropTypes.string,
-      hashed_email: PropTypes.string,
-    })).isRequired,
-    connect: PropTypes.func.isRequired,
-    disconnect: PropTypes.func.isRequired,
+    // Provided by redux
+    authenticationOptions: PropTypes.objectOf(authOptionPropType),
     userHasPassword: PropTypes.bool.isRequired,
     isGoogleClassroomStudent: PropTypes.bool.isRequired,
     isCleverStudent: PropTypes.bool.isRequired,
+    disconnect: PropTypes.func.isRequired,
   };
 
-  getAuthenticationOption = (provider) => {
-    return this.props.authenticationOptions.find(option => {
-      return option.credential_type === provider;
-    });
+  connect = (provider) => {
+    navigateToHref(`/users/auth/${provider}/connect`);
+  };
+
+  toggleProvider = (id, provider) => {
+    if (id) {
+      this.props.disconnect(id);
+    } else {
+      this.connect(provider);
+    }
+  };
+
+  cannotDisconnectGoogle = (authOption) => {
+    return authOption.credentialType === OAUTH_PROVIDERS.GOOGLE && this.props.isGoogleClassroomStudent;
   };
 
-  hasAuthOption = (provider) => {
-    return this.getAuthenticationOption(provider) !== undefined;
+  cannotDisconnectClever = (authOption) => {
+    return authOption.credentialType === OAUTH_PROVIDERS.CLEVER && this.props.isCleverStudent;
   };
 
-  getEmailForProvider = (provider) => {
-    const authOption = this.getAuthenticationOption(provider);
-    if (authOption) {
-      if (this.props.userType === 'student') {
-        return ENCRYPTED;
-      }
-      return authOption.email;
+  userHasLoginOption = (authOptions) => {
+    // It's the user's last authentication option
+    if (authOptions.length === 0) {
+      return false;
+    }
+
+    // If the user's only authentication option is an email address, a password is required for login
+    const otherOptionIsEmail = authOptions.length === 1 && authOptions[0].credentialType === 'email';
+    if (otherOptionIsEmail) {
```

Comment:
Nit: Although our UI doesn't support this, our data model allows a user to have multiple email authentication options.  Therefore, it would be more resilient to check here if _all_ remaining options are email, rather than checking for exactly one.

Reply:
good call! i'll make that change now